### ISSUE TITLE
Implement on top of new Generics method

### ISSUE THREAD

**Issue by ndmitchell** [ndmitchell/uniplate#11]
The new generics mechanism is up to 10x faster than SYB:

http://www.haskell.org/ghc/docs/latest/html/users_guide/generic-programming.html

I should reimplement Uniplate on top of that.

**Comment by minad** [comment]
You mean the GHC.Generics? Is Data.Data deprecated? And how much effort would it be to rewrite this to use GHC.Generics? I wonder what the recommended method is to rewrite ASTs (currently I use transformBi and transform).

**Comment by flip111** [comment]
10x faster! Sounds totally worth it. Here is a link to the latest doc page: https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#generic-programming

**Comment by ndmitchell** [comment]
I do mean GHC.Generics, and Data.Data isn't deprecated, I would leave both of them available. Rewriting ASTs with transformBi would just go 10x faster by magic (perhaps - perhaps not - you can never really tell in advance).

**Comment by minad** [comment]
Note that there is also geniplate, which is fast!

**Comment by jberryman** [comment]
@minad I saw that but was a little leery of using the fork package. Also I'm mostly interested in `transformBis` (plural); it's not clear to me whether composing several of these generated functions would be a performance win, since I assume ghc won't be able to optimize it into a single traversal

**Comment by minad** [comment]
@jberryman No experience with your special use case. The geniplate fork was broken from time to time, missing from stackage etc. But it is used by Agda, so it has always been fixed quickly. I would have to check, but I believe it is still used there.

**Comment by jberryman** [comment]
Right, just checked and they actually excised that dependency early this year, though they forgot to remove from stack.yaml

**Comment by minad** [comment]
> Right, just checked and they actually excised that dependency early this year, though they forgot to remove from stack.yaml

That's not good for me then :( Do you know what they are using instead? Generally I think most traversals should rather be written via functors, traversable and recursion-schemes where possible.

**Comment by ndmitchell** [comment]
The geniplate idea is a good one - use TemplateHaskell to guarantee zero-cost abstraction. Whether the implementation of geniplate is well maintained or not is harder to say - things that rely on TemplateHaskell require a lot of hand holding as the API changes frequently.

It still seems that investigating Uniplate on top of GHC Generics would be a good idea. I'm a little concerned about the compile times of GHC generics, but that's also true for lots of things in Haskell. A worthwhile addition to the library, if someone wants to take the time to do it.
